packages/odata-annotation-core/src/types/metadataService.ts

```diff
@@ -0,0 +1,362 @@
+type ActionNameMap = Map<Path, Set<Path>>;
+
+// OData conform target kinds for CDS kinds
+const targetKindsCds = {
+    service: [Edm.EntityContainer],
+    entitySet: [Edm.EntitySet],
+    entity: [Edm.EntityType, Edm.EntitySet],
+    view: [Edm.EntityType, Edm.EntitySet],
+    aspect: [Edm.ComplexType, Edm.EntityType, Edm.EntitySet],
+    element: [Edm.Property],
+    action: [Edm.Action, Edm.ActionImport],
+    function: [Edm.Function, Edm.FunctionImport],
+    actionImport: [Edm.Action],
+    functionImport: [Edm.FunctionImport],
+    param: [Edm.Parameter],
+    type: [Edm.TypeDefinition, Edm.Property, Edm.Parameter]
+};
+const targetKindsMapCds: Map<string, string[]> = new Map(Object.entries(targetKindsCds));
+
+/**
+ * Extracts action/function name by removing the part enclosed in parentheses.
+ *
+ * @param name action/function name
+ * @returns extracted name
+ */
+function getActionName(name: Path): Path {
+    const partsOpen = name.split('(');
+    const partsClose = name.split(')');
+    return partsOpen.length > 1 && partsClose.length > 1 ? (partsOpen.shift() ?? '') + (partsClose.pop() ?? '') : name;
+}
+
+/**
+ *
+ * @param metadata
+ * @param actionNames
+ * @param namespaces
+ * @param metadataElements
+ */
+function buildMetadataElementLookup(
+    metadata: MetadataMap,
+    actionNames: ActionNameMap,
+    namespaces: Set<string>,
+    metadataElements: MetadataElement[]
+): void {
+    for (const node of metadataElements) {
+        const namespace = getNamespace(node.path);
+        if (namespace) {
+            namespaces.add(namespace);
+        }
+        if (node.name.includes('(')) {
+            const actionName = getActionName(node.name);
+            let overloadPaths = actionNames.get(actionName);
+            if (!overloadPaths) {
+                overloadPaths = new Set<string>();
+                actionNames.set(actionName, overloadPaths);
+            }
+            overloadPaths.add(node.name);
+        }
+        metadata.set(node.path, node);
+        if (node.content) {
+            buildMetadataElementLookup(metadata, actionNames, namespaces, node.content);
+        }
+    }
+}
+
+/**
+ * Returns the namespace of an element identified by its path.
+ *
+ * @param path - Path of an element
+ * @returns    - Namespace
+ */
+function getNamespace(path: string): string {
+    let namespace = '';
+    const segments = path.split('/');
+    const firstSegment = segments[0] === '' ? segments[1] : segments[0];
+    if (firstSegment) {
+        // segment name for overloads can be <fqFunctionName>(<fqNamePar1>, <fqNamePar2>)
+        // each fqName can contain dots
+        const beforeFirstBracket = firstSegment.split('(')[0];
+        const parts = beforeFirstBracket.split('.');
+        parts.pop();
+        namespace = parts.join('.');
+    }
+    return namespace;
+}
+
+/**
+ * Metadata service
+ */
+export class MetadataService implements IMetadataService {
```

Comment:
We should not have the class here. It is types folder and the implementation should be in the metadata module.

Reply:
Modev to another unit, but anyway it is temporary location. When we have metadata service moved to oss repo from UX Tools, we will consume it from a separate package.